x-pack/plugin/async-search/qa/security/src/test/java/org/elasticsearch/xpack/search/AsyncSearchSecurityIT.java

```diff
@@ -57,9 +63,64 @@ public void indexDocuments() throws IOException {
 
         createIndex("index-user2", Settings.EMPTY);
         index("index-user2", "0", "foo", "bar");
+        index("index-user2", "1", "bar", "baz");
         refresh("index-user2");
     }
 
+    public void testWithDlsAndFls() throws Exception {
+        Response submitResp = submitAsyncSearch("*", "*", TimeValue.timeValueSeconds(10), "user_dls");
+        assertOK(submitResp);
+        Map<String, Object> submitRespMap = toMap(submitResp);
+        List<Map<String, Map<String, Object>>> hits = extractHits(submitRespMap);
+        assertThat(hits, contains(new BaseMatcher<>() {
+            @Override
+            @SuppressWarnings("unchecked")
+            public boolean matches(Object actual) {
+                if (actual instanceof Map) {
+                    Map<String, Object> searchHit = (Map<String, Object>) actual;
+                    return "index".equals(searchHit.get("_index")) &&
+                            "1".equals(searchHit.get("_id")) &&
+                            ((Map<String, Object>) searchHit.get("_source")).isEmpty();
+                }
+                return false;
+            }
```

Comment:
Speaking from personal preference, I think this is trying too hard to use a `Matcher`, and it would be more readable to use a different style.
For example, we know exactly what results are expected, so we can check the number of hits & then we can sort the hits and check that each element is what we expect it to be.

Also, I may be mistaken because all the casting of JSON is a bit hard to read, but I think you have a type mismatch here.

If `hits` is `List<Map<String, Map<String, Object>>>` 
then each element is `Map<String, Map<String, Object>>`  
but in the Matcher, you cast actual to `Map<String, Object>` and check that some of the values are strings.

I think hits needs to be `List<Map<String, ?>>` or `List<Map<String, Object>>`

Reply:
Okay, I have polished the test code here.
I've parsed the response to the actual `AsyncSearchResponse` so it doesn't deal with `Map`s and casts, but I've kept the collection matcher which now uses `CustomMatcher` elements which is simpler. I like this over sorting, because I don't know how to sort results across indices, and it feels contrived.